ci_feature/main.sh

```diff
@@ -186,26 +164,153 @@ echo "Making wget request to cadvisor endpoint with port 10250"
+         ITEMS_COUNT=$(echo $podsResponse | jq '.items | length')
+         echo "found items count: $ITEMS_COUNT"
+         if [ $ITEMS_COUNT -gt 0 ]; then 
+            # exclude the pods which doesnt have containerId. could happen if the container fails to start because of bad image and tag etc..
+            podsWithValidContainerId=$(echo $podsResponse | jq -r '[.items[] | select( .status.containerStatuses != null and .status.containerStatuses[].containerID != null and  .status.containerStatuses[].containerID != "")]')                         
+            ITEMS_COUNT_WITH_CONTAINER_ID=$(echo $podsWithValidContainerId | jq '. | length')
+            if [ $ITEMS_COUNT_WITH_CONTAINER_ID -gt 0 ]; then 
+                  # containerId name contract https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/#pod-v1-core
+                  containerRuntime=$(echo $podsWithValidContainerId | jq -r '.[0].status.containerStatuses[0].containerID' | cut -d ':' -f 1)                  
+                  nodeName=$(echo $podsWithValidContainerId | jq -r '.[0].spec.nodeName')
+                  # convert to lower case so that everywhere else can be used in lowercase
+                  containerRuntime=$(echo $containerRuntime | tr "[:upper:]" "[:lower:]")
+                  nodeName=$(echo $nodeName | tr "[:upper:]" "[:lower:]")
+                  # update runtime only if its not empty and not startswith docker
+                  if [ -z $containerRuntime ]; then
+                      echo "using default container runtime as docker since got containeRuntime as empty string"                              
+                  elif [[ $containerRuntime != docker* ]]; then                     
+                       export CONTAINER_RUNTIME=$containerRuntime                                          
+                  fi
+
+                  if [ -z $nodeName ]; then
+                      echo "-e error nodeName in /pods API response is empty"
+                  else
+                     export NODE_NAME=$nodeName   
+                  fi                   
+            else
+              echo "-e error  none of the pods in the /pods response has valid containerID"                           
+            fi   
+        else
+             echo "-e error  items in the /pods response is 0"                           
+        fi      
+      else
+          echo "-e error  wget request to cadvisor endpoint /pods with port 10250 to get the configured container runtime on kubelet failed"    
+      fi
+
 else
       echo "Wget request using port 10250 failed. Using port 10255"
       export IS_SECURE_CADVISOR_PORT=false
       echo "export IS_SECURE_CADVISOR_PORT=false" >> ~/.bashrc
       export CADVISOR_METRICS_URL="http://$NODE_IP:10255/metrics"
       echo "export CADVISOR_METRICS_URL=http://$NODE_IP:10255/metrics" >> ~/.bashrc
+      echo "Making wget request to cadvisor endpoint with port 10255 to get the configured container runtime on kubelet"
```

Comment:
Why do we have 2 copies of the same code for secure and insecure? I think it should be extracted to a method to make it more readable and remove duplication.

Reply:
> Why do we have 2 copies of the same code for secure and insecure? I think it should be extracted to a method to make it more readable and remove duplication.

> one for secure and other one for unsecure. sure, at certain point we can refactor this.

Reply:
Why not now? :) It is a huge block of code that is duplicated. It would be much cleaner to just have one copy of the code since it is in the same file.

Reply:
> Why not now? :) It is a huge block of code that is duplicated. It would be much cleaner to just have one copy of the code since it is in the same file.

Done. :).

Reply:
Great! Thanks :)